### PR TITLE
Do not try to read xcodebuild log file if it is too big

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
See https://github.com/appium/appium/issues/9256 for more details

### CODE REVIEW

**Comment by imurchie** [comment]
Alternatively this check could be changed to just running an external grep command on the file.

**Comment by mykola-mokhnach** [comment]
ok, let's grep it

File: lib/wda/xcodebuild.js [hunk 1]
```diff
@@ -16,6 +16,7 @@ const BUILD_TEST_DELAY = 1000;
 
 const DERIVED_DATA_FOLDER_REGEXP = /(\/.+\/DerivedData\/WebDriverAgent-[^\/]+)/;
 const DERIVED_DATA_LOG_REGEXP = /\s+(\/.+\/WebDriverAgentRunner-.+\/.+\.log)/;
+const DERIVED_DATA_GREP_REGEXP = '/WebDriverAgentRunner-';
```
Comment (imurchie) [inline lib/wda/xcodebuild.js]: If this is called a regexp it ought to be a regexp.

File: lib/wda/xcodebuild.js [hunk 2]
```diff
@@ -106,11 +107,17 @@ class XcodeBuild {
...
       // now parse the log file for the derived data folder
       const logFile = match[1];
-      stdout = await fs.readFile(logFile);
-      match = DERIVED_DATA_FOLDER_REGEXP.exec(stdout);
-      if (!match) {
-        // nothing found. We are done
-        log.debug(`Cannot find the derived data location from the xcodebuild log file '${logFile}'`);
+      try {
+        // grep on the log file, since it might be too big to fit it completely into the memory
+        const grepData = await exec('grep', DERIVED_DATA_GREP_REGEXP, logFile);
```
Comment (imurchie) [inline lib/wda/xcodebuild.js]: `exec` takes a command and an array of arguments to that command. https://github.com/appium/node-teen_process/blob/master/lib/teen_process.js#L11

Comment (mykola-mokhnach) [reply lib/wda/xcodebuild.js]: fixed

File: lib/wda/xcodebuild.js (hunk 1 above)
Comment (mykola-mokhnach) [reply lib/wda/xcodebuild.js]: it's not a real regular expression object, since it's just used as a grep argument. I've renamed the constant

**Comment by cosminstirbu** [comment]
Hi,

I think I am having a similar problem.

After running my iOS tests for approx 40mins, I receive the following error when creating a new driver:

```
[XCUITest] Error: "toString()" failed
    at Buffer.toString (buffer.js:513:11)
    at RegExp.exec (native)
    at XcodeBuild.retrieveDerivedDataPath$ (../../../lib/wda/xcodebuild.js:110:42)
    at tryCatch (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:67:40)
    at GeneratorFunctionPrototype.invoke [as _invoke] (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:315:22)
    at GeneratorFunctionPrototype.prototype.(anonymous function) [as next] (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:100:21)
    at GeneratorFunctionPrototype.invoke (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:136:37)
 Error: "toString()" failed
    at Buffer.toString (buffer.js:513:11)
    at RegExp.exec (native)
    at XcodeBuild.retrieveDerivedDataPath$ (../../../lib/wda/xcodebuild.js:110:42)
    at tryCatch (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:67:40)
    at GeneratorFunctionPrototype.invoke [as _invoke] (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:315:22)
    at GeneratorFunctionPrototype.prototype.(anonymous function) [as next] (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:100:21)
    at GeneratorFunctionPrototype.invoke (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:136:37)
[MJSONWP] Encountered internal error running command: Error: "toString()" failed
    at Buffer.toString (buffer.js:513:11)
    at RegExp.exec (native)
    at XcodeBuild.retrieveDerivedDataPath$ (../../../lib/wda/xcodebuild.js:110:42)
    at tryCatch (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:67:40)
    at GeneratorFunctionPrototype.invoke [as _invoke] (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:315:22)
    at GeneratorFunctionPrototype.prototype.(anonymous function) [as next] (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:100:21)
    at GeneratorFunctionPrototype.invoke (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:136:37)
[31mFailure in after hook:[0m[31mHooks.onScenarioEnded(Scenario)[0m
[31mMessage: [0m[31morg.openqa.selenium.WebDriverException: An unknown server-side error occurred while processing the command. Original error: "toString()" failed (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 31.61 seconds
Build info: version: '2.53.0', revision: '35ae25b1534ae328c771e0856c93e187490ca824', time: '2016-03-15 10:43:46'
System info: host: 'osxbuildserver.internal.mywebgrocer.com', ip: '10.0.40.9', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.12.6', java.version: '1.8.0_91'
Driver info: io.appium.java_client.ios.IOSDriver
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.openqa.selenium.remote.ErrorHandler.createThrowable(ErrorHandler.java:206)
	at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed(ErrorHandler.java:158)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:678)
	at io.appium.java_client.DefaultGenericMobileDriver.execute(DefaultGenericMobileDriver.java:40)
	at io.appium.java_client.AppiumDriver.execute(AppiumDriver.java:1)
	at io.appium.java_client.ios.IOSDriver.execute(IOSDriver.java:1)
	at org.openqa.selenium.remote.RemoteWebDriver.startSession(RemoteWebDriver.java:249)
	at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:131)
	at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:144)
	at io.appium.java_client.DefaultGenericMobileDriver.<init>(DefaultGenericMobileDriver.java:36)
	at io.appium.java_client.AppiumDriver.<init>(AppiumDriver.java:114)
	at io.appium.java_client.AppiumDriver.<init>(AppiumDriver.java:132)
	at io.appium.java_client.ios.IOSDriver.<init>(IOSDriver.java:82)
	at utils.TestBase.startDriver(TestBase.java:93)
	at utils.TestBase.restartService(TestBase.java:58)
	at glue.hooks.Hooks.onScenarioEnded(Hooks.java:28)
	at sun.reflect.GeneratedMethodAccessor41.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at cucumber.runtime.Utils$1.call(Utils.java:40)
	at cucumber.runtime.Timeout.timeout(Timeout.java:16)
	at cucumber.runtime.Utils.invoke(Utils.java:34)
	at cucumber.runtime.java.JavaHookDefinition.execute(JavaHookDefinition.java:60)
	at cucumber.runtime.Runtime.runHookIfTagsMatch(Runtime.java:224)
	at cucumber.runtime.Runtime.runHooks(Runtime.java:212)
	at cucumber.runtime.Runtime.runAfterHooks(Runtime.java:206)
	at cucumber.runtime.model.CucumberScenario.run(CucumberScenario.java:46)
	at cucumber.runtime.junit.ExecutionUnitRunner.run(ExecutionUnitRunner.java:102)
	at cucumber.runtime.junit.FeatureRunner.runChild(FeatureRunner.java:63)
	at cucumber.runtime.junit.FeatureRunner.runChild(FeatureRunner.java:18)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at cucumber.runtime.junit.FeatureRunner.run(FeatureRunner.java:70)
	at cucumber.api.junit.Cucumber.runChild(Cucumber.java:95)
	at cucumber.api.junit.Cucumber.runChild(Cucumber.java:38)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at cucumber.api.junit.Cucumber.run(Cucumber.java:100)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)
```

Will this PR fix the issue?

Thank you,
Cosmin

**Comment by cosminstirbu** [comment]
Also, is there a way to bring these changes into my Appium 1.7.0 installation?
I am not too familiar with the Node modules deployment system, not sure if I can pull the appium-xcuitest-driver dependency separately from the master branch :-?

**Pull request merged**